worker/src/handlers.rs

```diff
@@ -1,13 +1,163 @@
+            Option<usize>,
+            WorkerBatchRequest,
+            oneshot::Sender<Result<anemo::Response<WorkerBatchResponse>>>,
+        )>,
+        rx_reconfigure: watch::Receiver<ReconfigureNotification>,
+        network: P2pNetwork,
+    ) -> JoinHandle<()> {
+        tokio::spawn(async move {
+            Self {
+                name,
+                id,
+                worker_cache,
+                request_batches_timeout,
+                rx_request_batches_rpc,
+                rx_reconfigure,
+                network,
+            }
+            .run()
+            .await;
+        })
+    }
+
+    async fn handle_worker_batch_responses(
+        responses: Vec<Timeout<JoinHandle<Result<anemo::Response<WorkerBatchResponse>>>>>,
+        response_channel: oneshot::Sender<Result<anemo::Response<WorkerBatchResponse>>>,
+    ) {
+        let mut results: FuturesUnordered<_> = responses.into_iter().collect();
+        let mut last_result: Result<anemo::Response<WorkerBatchResponse>> = Err(anyhow::anyhow!(
+            "no WorkerBatchResponse received (or request(s) timed out)"
+        ));
+        while let Some(result) = results.next().await {
+            match result {
+                Ok(Ok(Ok(result))) => {
+                    let _ = response_channel.send(Ok(result));
+                    return;
+                }
+                Ok(Ok(Err(e))) => last_result = Err(e),
+                _ => (),
+            }
+        }
+        let _ = response_channel.send(last_result);
+    }
+
+    async fn run(&mut self) {
+        let mut inflight_requests = FuturesUnordered::new();
+        loop {
+            tokio::select! {
+                Some((target, num_nodes, request, response_channel)) = self.rx_request_batches_rpc.recv() => {
+                    if target.is_some() && num_nodes.is_some() {
+                        panic!("cannot set both target and num_nodes");
+                    }
+                    if let Some(target) = target {
+                        match self.network.unreliable_send(target, &request) {
+                            Ok(handle) => {
+                                inflight_requests.push(Self::handle_worker_batch_responses(
+                                    vec![time::timeout(self.request_batches_timeout, handle)], response_channel))
+                            },
+                            Err(e) => {
+                                let _ = response_channel.send(Err(e));
+                            },
+                        }
+                    } else if let Some(num_nodes) = num_nodes {
+                        let names = self.worker_cache.load()
+                            .others_workers(&self.name, &self.id)
+                            .into_iter()
+                            .map(|(_, info)| info.name)
+                            .collect();
+                        let handles: Vec<_> = self.network.lucky_broadcast(names, &request, num_nodes)
```

Comment:
Do we want to give an int `num_nodes`, or a list of them (since we usually request a batch given a cert with a list of signers).

Reply:
This preserves the existing behavior from synchronizer, which chooses randomly among all other workers. I tried to keep same behavior as much as possible since this PR is already huge; hoping it'll be easier to make additional business logic tweaks separate from the big refactoring of the call pattern